File: src/main/java/algo/backtrack/NonDecreasingSubsequences_491.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class NonDecreasingSubsequences_491 {
    public static List<List<Integer>> findSubsequences(int[] nums) {
        var sequences = new HashSet<List<Integer>>();
        backtrack(nums, 0, new ArrayList<>(), sequences);
        return sequences.stream().toList();
    }

    private static void backtrack(int[] nums, int start, List<Integer> sub, Set<List<Integer>> all) {
        if (sub.size() > 1) {
              all.add(new ArrayList<>(sub));
        }
        for (var i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            if (sub.isEmpty() || sub.getLast() <= nums[i]) {
                sub.add(nums[i]);
                backtrack(nums, i + 1, sub, all);
                sub.remove(sub.size() - 1);
            }
        }
    }

    public static void main(String[] args) {
//        var seq = findSubsequences(new int[]{4,6,7,7});
//        var seq = findSubsequences(new int[]{4,4,3,2,1});
//        var seq = findSubsequences(new int[]{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15});
        var seq = findSubsequences(new int[]{1,2,3,4,5,6,7,8,9,10,1,1,1,1,1});
        seq.forEach(System.out::println);
    }
}

File: src/main/kotlin/algo/backtrack/NonDecreasingSubsequences_491.kt
package algo.backtrack

class KtNonDecreasingSubsequences_491 {
    fun findSubsequences(nums: IntArray): List<List<Int>> {
        val all: MutableSet<List<Int>> = mutableSetOf()
        fun backtrack(nums: IntArray, start: Int, list: MutableList<Int>) {
            if (list.size > 1) {
                all.add(list.toList())
            }
            for (i in start..<nums.size) {
                if (i > start && nums[i] == nums[i - 1]) {
                    continue
                }
                if (list.isEmpty() || nums[i] >= list.last()) {
                    list.add(nums[i])
                    backtrack(nums, i + 1, list)
                    list.removeLast()
                }
            }
        }
        backtrack(nums, 0, mutableListOf())
        return all.toList()
    }
}

fun main() {
    val sub = KtNonDecreasingSubsequences_491()
    println(sub.findSubsequences(intArrayOf(4,6,7,7)))
}